lib/util/graceful_shutdown.c

```diff
@@ -0,0 +1,101 @@
+#include <signal.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+/* Data from parent code. */
+static int (* begin_shutdown)(void *);
+static void (* sighandler_sigterm_orig)(int);
+static void * caller_cookie;
+static void * timer_cookie;
+
+/* Flag to show that SIGTERM was given. */
```

Comment:
I'd argue that we don't care about *receiving* SIGTERM; what we care about was that it was sent to us.  Which implicitly means that we received it, but ultimately that's not what we care about.  However, I'd much rather get this merged, so I'll refrain from making that argument.  :)